pkg/services/ngalert/state/manager.go

```diff
@@ -247,7 +252,11 @@ func (st *Manager) DeleteStateByRuleUID(ctx context.Context, ruleKey ngModels.Al
 		s.SetNormal(reason, startsAt, now)
 		// Set Resolved property so the scheduler knows to send a postable alert
 		// to Alertmanager.
-		s.Resolved = oldState == eval.Alerting || oldState == eval.Error || oldState == eval.NoData
+		if oldState == eval.Alerting || oldState == eval.Error || oldState == eval.NoData {
+			s.ResolvedAt = now
```

Comment:
What about `pending`? Let's imagine a situation when:
`Normal -> Pending -> Alerting -> Normal -> Pending -> Normal`

In the case of the second and thrid `Normal` we won't send the resolved notification.

Reply:
The 2nd Normal will always send because we now check if `a.ResolvedAt.After(a.LastSentAt)` in `NeedsSending(...)`.

The 3rd Normal I'm on the fence about. Currently, I reset `ResolvedAt` when entering Pending as this follows how Prom Ruler does this. However, this isn't exactly obvious IMO and had to look it up to see what they decided to do here.

I lean slightly towards not resetting `ResolvedAt` when entering Pending, but I'd have to give some through to situations where this might have unintended consequences.